chain/chain.go

```diff
@@ -0,0 +1,270 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chain
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/api/metrics"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/trace"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/hypersdk/internal/emap"
+	"github.com/ava-labs/hypersdk/internal/workers"
+	"github.com/ava-labs/hypersdk/state"
+
+	internalfees "github.com/ava-labs/hypersdk/internal/fees"
+)
+
+type Chain struct {
```

Comment:
I feel like `Chain` is a weird abstraction to me - it's not clear to me what it's responsibility is and it feels like it's turning into a [god-object](https://en.wikipedia.org/wiki/God_object).

Reply:
This is effectively replacing the static functions from the chain package that were used before.

It's not perfect, but much better than the previous code imo. The end goal is to decouple the VM from the block, state, and result types completely.

We could separate the builder/parser/accepter etc, but they should all have the same block/state/result types.

Reply:
Separated them out into individual components, but kept a single chain type to wrap them together